src/handlers/psp.ts

```diff
@@ -0,0 +1,436 @@
+    const { airnodeWallet, providerName, chainId, provider, contracts, voidSigner, currentBlock, gasTarget } =
+      providerState;
+
+    const providerLogOptions: node.LogOptions = {
+      ...baseLogOptions,
+      meta: {
+        ...baseLogOptions.meta,
+        chainId,
+        providerName,
+      },
+    };
+
+    // Get subscriptions from template/endnpoint groups
+    const subscriptions = flatMap(groupedSubscriptions.map((s) => s.subscriptions));
+
+    // Filter subscription by chainId and doblue-check that subscription has an associated API value
+    const chainSubscriptions = subscriptions.filter(
+      (subscription) => subscription.chainId === chainId && apiValuesBySubscriptionId[subscription.id]
+    );
+
+    // Check conditions
+    const validSubscriptions = await checkSubscriptionsConditions(
+      chainSubscriptions,
+      apiValuesBySubscriptionId,
+      contracts['DapiServer'],
+      voidSigner,
+      providerLogOptions
+    );
+
+    // Group subscriptions by sponsor address
+    const subscriptionsBySponsor = groupBy(validSubscriptions, 'sponsor');
+
+    // Fetch sponsor wallet transaction counts to be able to assign nonces
+    // to subscriptions and group subscriptions by sponsor wallet
+    const subscriptionsBySponsorWallets = await groupSubscriptionsBySponsorWallet(
+      subscriptionsBySponsor,
+      config,
+      provider,
+      currentBlock,
+      providerLogOptions
+    );
+
+    // Process sponsor wallets in parallel
+    const sponsorWalletPromises = subscriptionsBySponsorWallets.map(async ({ subscriptions, sponsorWallet }) => {
+      const shortSponsorWalletAddress = sponsorWallet.address.replace(sponsorWallet.address.substring(5, 38), '...');
+
+      const sponsorWalletLogOptions: node.LogOptions = {
+        ...providerLogOptions,
+        additional: {
+          ...providerLogOptions.additional,
+          sponsorWallet: shortSponsorWalletAddress,
+        },
+      };
+
+      node.logger.info(`Processing ${subscriptions.length} subscription(s)`, sponsorWalletLogOptions);
+
+      const logs = await processTransaction(
+        airnodeWallet,
+        contracts['DapiServer'],
+        gasTarget,
+        subscriptions,
+        sponsorWallet
+      );
+
+      logs.forEach(([logs, data]) => {
+        const subscriptionLogOptions: node.LogOptions = {
+          ...sponsorWalletLogOptions,
+          additional: {
+            ...sponsorWalletLogOptions.additional,
+            subscriptionId: data.id,
+          },
+        };
+        node.logger.logPending(logs, subscriptionLogOptions);
+      });
+    });
+
+    await Promise.all(sponsorWalletPromises);
+  });
+
+  await Promise.all(providerPromises);
+};
+
+const updateBeacon = async (config: Config) => {
+  // =================================================================
+  // STEP 1: Initialize state
+  // =================================================================
+  let state: State = initializeState(config);
```

Comment:
;)

Reply:
Should I just rename this function  to `state()` and pass an arg called `state` 🤣